tensorboard/data/server/run.rs

```diff
@@ -69,6 +66,20 @@ enum EventFile<R> {
     Dead,
 }
 
+/// Holds data staged by a RunLoader that will be commited to the Commit.
+#[derive(Debug)]
+struct RunLoaderData {
```

Comment:
I like the idea of extracting this type.

When first writing this module, I considered separating `RunLoader` into
`files` and `data` and putting methods on `data`, to cut down on
free-floating functions taking a bunch of `&mut` arguments. It seems
like that’s about what you’re doing here, so, some observations:

  - You’ve written this semantically as “the data that will be
    committed”, which makes sense. You could also write it more
    structurally as “the data that isn’t in `files`, since that’s the
    field into which we like to take long-lived mutable references”.
    I.e.: you could include `self.run` and `self.checksum` here.

    Doing so avoids the clone of `run_name`, which obviously isn’t a big
    performance hit, but makes me wonder whether the structure might be
    nice. It also simplifies the callback structure a bit (see below).

  - The free-standing `fn commit_all` and `fn read_event` functions are
    nicely set up to be `&mut self` methods on `RunLoaderData`, which I
    viewed as one of the benefits of this structure in the first place.

Thoughts?

Reply:
Good point about how the functions could be methods, I hadn't thought that far.  Can do.

Re: the scope of the data structure, I did briefly think about which fields to include and it seemed like a more semantically coherent boundary to define it as the data that will be committed.  Defining as just "everything except files" felt odd to me even if functionally it's the most convenient.  In particular even though cloning `run` is a bit ugly, semantically it just feels quite weird to make `run` a field on `self.data` instead of the top-level `RunLoader`, so I guess I preferred putting the weirdness inside the method bodies versus into the struct layout.  Similarly, it seems like semantically `self.checksum` would be more naturally located alongside `self.files` since it only affects the file-reading half of `RunLoader`, not the data-buffering half.

I guess in the purest sense we want something like a tripartite structure in which we have "effectively constant metadata" like `self.run` and `self.checksum` grouped in one substructure, and and then two different mutable substructures, `self.files` and `self.data`, so that all three can be reborrowed independently from the parent struct, but that seems... overcomplicated?  Would be nice to have some way to help the borrow checker resolve this without as much reshuffling of the actual struct fields themselves.

Thoughts?

Reply:
Yep, that makes sense and seems reasonable to me. No strong opinion;
feel free to proceed as desired.

> Would be nice to have some way to help the borrow checker resolve this without as much reshuffling of the actual struct fields themselves.

You may be interested in [@nikomatsakis’s article on interprocedural
conflicts][1]. (Context: “NLL” = “non-lexical lifetimes”, referring to
a set of changes to the borrow checker taught it to accept more kinds of
valid code, making it much easier to write programs that compile.)

[1]: https://smallcultfollowing.com/babysteps/blog/2018/11/01/after-nll-interprocedural-conflicts/

Reply:
Converted the free functions into methods.

Yeah, that article pretty much sums up the understanding I had come to as well.  The "View" thing is interesting, I could see that being fairly effective for a more complex situation but it definitely isn't lightweight.  Thanks for sharing!